File: 2022/day-7/part-1.py
# Copyright (c) 2022 David Chan
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT

class Directory:
    def __init__(self, name, parent=None):
        self.name = name
        self.children = []
        self.files = []
        self.parent = parent

    def size(self):
        return sum([file[1] for file in self.files]) + sum([child.size() for child in self.children])


def get_small_dirs(root):
    small_dirs = []
    for child in root.children:
        if child.size() < 100000:
            small_dirs.append(child)
        small_dirs += get_small_dirs(child)
    return small_dirs


def main():

    # Load the input file
    with open("input.txt", "r") as f:
        lines = f.readlines()

    # Create the root directory
    root = Directory("/")
    current_dir = root

    for line in lines:
        if line.startswith("$"):
            # This is a command line
            tok = line.split()
            if tok[1] == "cd":
                # Change directory
                if tok[2] == "..":
                    # Go up one directory
                    current_dir = current_dir.parent or current_dir
                elif tok[2] == '/':
                    # Go to the root directory
                    current_dir = root
                else:
                    for child in current_dir.children:
                        if child.name == tok[2]:
                            current_dir = child
                            break
        else: # We're in LS mode --
            # This is a file or directory
            size, name = line.split()
            if size == 'dir':
                if name not in [child.name for child in current_dir.children]:
                    # This is a directory
                    new_dir = Directory(name, current_dir)
                    current_dir.children.append(new_dir)
            else:
                if name not in [file[0] for file in current_dir.files]:
                    # This is a file
                    current_dir.files.append((name, int(size)))

    print('Total Size', root.size())

    # Get every directory with size < 100000
    small_dirs = get_small_dirs(root)

    # Print the names of the small directories
    for small_dir in small_dirs:
        print(small_dir.name)

    # Get the sum of the sizes of the small dirs
    print('Sum of small dirs', sum([small_dir.size() for small_dir in small_dirs]))


if __name__ == "__main__":
    main()

File: 2022/day-7/part-1.rs
// Copyright (c) 2022 David Chan
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

use std::env;
use std::fs;

struct FileSystem {
    name: String,
    directories: Vec<usize>,
    files: Vec<(String, usize)>,
    parent: usize,
}

impl FileSystem {
    pub fn new(name: String, parent: usize) -> FileSystem {
        FileSystem {
            name,
            directories: Vec::new(),
            files: Vec::new(),
            parent,
        }
    }
}

fn get_directory_size(filesystems: &Vec<FileSystem>, index: usize) -> usize {
    let mut size = 0;
    for (_, file_size) in &filesystems[index].files {
        size += file_size;
    }
    for directory in &filesystems[index].directories {
        size += get_directory_size(filesystems, *directory);
    }
    size
}

fn main() {
    let filename = env::args().nth(1).expect("No filename given");
    let contents = fs::read_to_string(filename).expect("Could not read file");

    // Get each of the lines
    let lines: Vec<&str> = contents.split("\n").filter(|line| line.len() > 0).collect();

    let mut filesystems = Vec::new();

    // Add the root filesystem
    filesystems.push(FileSystem::new("/".to_string(), 0));
    let mut cwd = 0;

    // Parse each line
    let mut line_num = 0;
    for line in lines {
        // Command is "$ cd" or "$ ls" or a file descriptor
        if line.starts_with("$") {
            // Get the command
            let command = line.split(" ").nth(1).unwrap();
            match command {
                "cd" => {
                    // Get the directory name
                    let directory = line.split(" ").nth(2).unwrap();
                    if directory.eq("..") {
                        // Go up a directory
                        cwd = filesystems[cwd].parent;
                        continue;
                    } else if directory.eq("/") {
                        cwd = 0;
                    } else {
                        // Find the directory in the current working directory
                        let index = filesystems[cwd]
                            .directories
                            .iter()
                            .position(|&x| filesystems[x].name.eq(directory))
                            .unwrap();
                        // Set the current working directory to the index of the directory
                        cwd = filesystems[cwd].directories[index];
                    }
                }
                "ls" => {
                    // Ignore this - since it doesn't matter.
                }
                _ => panic!("Unknown command"),
            }
        } else {
            // File descriptor
            let (size, name) = line.split_once(" ").unwrap();
            if size.eq("dir") {
                // Check to see if the directory already exists
                if filesystems[cwd]
                    .directories
                    .iter()
                    .any(|&x| filesystems[x].name.eq(name))
                {
                    continue;
                }
                // Create a new directory
                filesystems.push(FileSystem::new(name.to_string(), cwd));
                // Add the index of the directory to the current working directory
                let index = filesystems.len() - 1;
                filesystems[cwd].directories.push(index);
            } else {
                // Check to see if the file already exists
                if filesystems[cwd].files.iter().any(|x| x.0.eq(name)) {
                    continue;
                }
                // Add the file to the current working directory
                // Print the file
                filesystems[cwd]
                    .files
                    .push((name.to_string(), size.parse::<usize>().unwrap()));
            }
        }
        line_num += 1;
    }

    // Get all of the directories with size < 100000
    let mut small_dirs = Vec::new();
    for (i, filesystem) in filesystems.iter().enumerate() {
        let size = get_directory_size(&filesystems, i);
        if size < 100000 {
            small_dirs.push((i, size));
        }
    }

    // Print the sum of the sizes of all of the directories with size < 100000
    let mut total = 0;
    for (_, size) in small_dirs {
        total += size;
    }

    println!("{}", total);
}
